From 0e841cf57d1c329d290a0bfcfa2398c3ced0fe39 Mon Sep 17 00:00:00 2001
From: Dan Villiom Podlaski Christiansen <danchr@gmail.com>
Date: Wed, 21 Oct 2020 16:48:38 +0200
Subject: [PATCH] Fix Autoconf checks on Darwin

The Autoconf checks merely verify that the symbol is available and
linkable, but not that it's actually defined and usable. This breaks
compilation on recent versions of macOS, where `lgamma_r()` is
unavailable unless `_REENTRANT` is defined.

While at it, I added support for `exp10()` on macOS, which is
available, but named `__exp10` due to its nonstandard nature.
---
 config/m4/check-math-func.m4 |  6 ++++--
 configure.ac                 | 20 ++++++++++++++++++++
 2 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/config/m4/check-math-func.m4 b/config/m4/check-math-func.m4
index 5677cede..a8aee81f 100644
--- a/config/m4/check-math-func.m4
+++ b/config/m4/check-math-func.m4
@@ -2,8 +2,10 @@ dnl AC_CHECK_MATH_FUNC(func)
 AC_DEFUN([AC_CHECK_MATH_FUNC], [
   AC_LANG(C)
   AC_CHECK_LIB([m],[$1],[
-    eval "ac_tr_func=HAVE_[]upcase($1)"
-    AC_DEFINE_UNQUOTED($ac_tr_func)
+    AC_CHECK_DECL([$1], [
+      eval "ac_tr_func=HAVE_[]upcase($1)"
+      AC_DEFINE_UNQUOTED($ac_tr_func)
+    ],[], [[#include <math.h>]])
   ],[
   ])
 ])
diff --git a/configure.ac b/configure.ac
index 0441d4a2..0c5056cc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -214,6 +214,26 @@ AC_CHECK_MATH_FUNC(y0)
 AC_CHECK_MATH_FUNC(y1)
 AC_CHECK_MATH_FUNC(yn)
 
+# exp10 is nonstandard, but available as __exp10 on Darwin
+if test $ac_cv_lib_m_exp10 = no; then
+  AC_CHECK_DECL([__exp10], [
+    AC_DEFINE_UNQUOTED([HAVE_EXP10])
+    AC_DEFINE_UNQUOTED([exp10], [__exp10])
+  ],[], [[#include <math.h>]])
+fi
+
+# check whether _REENTRANT is required for lgamma_r
+if test "x$ac_cv_have_decl_lgamma_r" = xno; then
+  unset ac_cv_have_decl_lgamma_r
+  AC_CHECK_DECL([lgamma_r], [
+    AC_DEFINE_UNQUOTED([_REENTRANT])
+    AC_DEFINE_UNQUOTED([HAVE_LGAMMA_R])
+  ],[], [[
+    #define _REENTRANT
+    #include <math.h>
+  ]])
+fi
+
 dnl Thread local storage
 have___thread=no
 AC_MSG_CHECKING(for thread-local storage)
